### ISSUE TITLE
Muskingum routing

### ISSUE THREAD

**Issue by tyralla** [hydpy-dev/hydpy#85]
So far, *HydPy* implements only a very simple Muskingum-like routing approach:  [hstream_v1](https://hydpy-dev.github.io/hydpy/4.0/hstream_v1.html).  It follows *HBV96* and comes with two parameters, [Lag](https://hydpy-dev.github.io/hydpy/4.0/hstream.html#hydpy.models.hstream.hstream_control.Lag) and [Damp](https://hydpy-dev.github.io/hydpy/4.0/hstream.html#hydpy.models.hstream.hstream_control.Damp), which describe the translation and diffusion of flood waves, respectively.  We want to add a more advanced alternative and decided on a variant of the [Muskingum-Cunge method modified by Todini (MCT)](https://hess.copernicus.org/articles/11/1645/2007/).

We want to introduce a new model family for it, probably named *HydPy-Musk*, for now only including [hstream_v1](https://hydpy-dev.github.io/hydpy/4.0/hstream_v1.html) and the new MCT model.  We can add other routing approaches similar to the Muskingum method later.

We decided against a general "routing model family" due to the different nature of approaches already implemented.  [arma_v1](https://hydpy-dev.github.io/hydpy/4.0/arma_v1.html) is an (extended) Unit-Hydrograph method and [lstream_v001](https://hydpy-dev.github.io/hydpy/4.0/lstream_v001.html) is a set of ordinary differential equations that need to be solved numerically.  Even when putting them all in a single model family, there are few functionalities and concepts they could share.  A short description for *HydPy-Musk* could be "routing methods that rely on non-adaptive finite difference approximations".

When moving [hstream_v1](https://hydpy-dev.github.io/hydpy/4.0/hstream_v1.html), we need to rename it.  In current terminology, the new name would be `musk_v001`.  However, we tend to make *HydPy-Musk* the first model family with more self-explaining model names.  Therefore, we need to clarify its scope more carefully.

[hstream_v1](https://hydpy-dev.github.io/hydpy/4.0/hstream_v1.html) applies the [standard Muskingum formula](https://hydpy-dev.github.io/hydpy/4.0/hstream.html#hydpy.models.hstream.hstream_model.Calc_QJoints_V1) with fixed coefficients on several channel segments.  The calculation of these three coefficients happens during model initialisation based on the current values of  [Lag](https://hydpy-dev.github.io/hydpy/4.0/hstream.html#hydpy.models.hstream.hstream_control.Lag) and [Damp](https://hydpy-dev.github.io/hydpy/4.0/hstream.html#hydpy.models.hstream.hstream_control.Damp).  The standard Muskingum method does not divide a channel into segments and uses the parameters `x` (related to the diffusion) and `k` (related to the translation time) for calculating the coefficients.

I suggest the following: we reformulate [hstream_v1](https://hydpy-dev.github.io/hydpy/4.0/hstream_v1.html) with two new parameters.   The first one is `NmbSegments`, which roughly corresponds to the old parameter [Lag](https://hydpy-dev.github.io/hydpy/4.0/hstream.html#hydpy.models.hstream.hstream_control.Lag).  The second parameter is a flexible "meta-parameter" that allows defining the coefficients of the working equation both in the standard Muskingum and the HBV96 way.  

A first draft on how to configure the adjusted model according to the standard Muskingum method:

```python
nmbseqments(1)
metaparameter(x=0.2, k=1000.0)
```

A first draft on how to configure the adjusted model according to the HBV96:

```python
nmbseqments(2)
metaparameter(damp=0.5)
```
`MetaParameter` is not an overly helpful name.   Maybe someone else can think of something more concrete?

In contrast to Muskingum-Cunge, the "classic" Muskingum method is time-invariant.  Maybe we could include this information in the new model name.  My first ideas:

- *musk_time_invariant*
- *musk_invariant*
- *musk_fixed_coeffs*

Or, maybe a less academic name:

- *musk_classic*
- *musk_simple*
- *musk_base*

Again, better ideas are welcome.

**Comment by BGWKlein** [comment]
use name nmbsegments instead of nmbseqments

Probably use also the HBV parameter lag [should always be in days] similar to damp as argument to derive the nmbsegments

nmbsegments(lag=2)
metaparameter(damp=0.5)

**Comment by tyralla** [comment]
> use name nmbsegments instead of nmbseqments

Seems like I write "sequences" too often...

> Probably use also the HBV parameter lag [should always be in days]...

Yes, this is a good idea.  But I would prefer to make the time unit of the *lag* argument dependent on the current value of the [parameterstep](https://hydpy-dev.github.io/hydpy/4.0/optiontools.html#hydpy.core.optiontools.Options.parameterstep) option.  Maybe this is less consistent with *HBV96*  but more consistent with how we usually set parameter value in *HydPy* control files.

>MetaParameter is not an overly helpful name.  Maybe someone else can think of something more concrete?

I still cannot come up with a better idea. Maybe just `Coefficients`.  Then, we could also allow to directly define all three coefficients of the working equation:

```python
coefficients(c1=0.2, c2=0.5, c3=0.3)
```

**Comment by tyralla** [comment]
We decided on [musk_classic ](https://hydpy-dev.github.io/hydpy/master/musk_classic.html) and implemented it in commit a77877511fe3f11b06f1914ea4b5e4186e71071d, which is already in master.  Unfortunately, implementing the "keyword-based" value definitions of parameters [NmbSegments](https://hydpy-dev.github.io/hydpy/master/musk.html#hydpy.models.musk.musk_control.NmbSegments) and [Coefficients](https://hydpy-dev.github.io/hydpy/master/musk.html#hydpy.models.musk.musk_control.Coefficients) introduced some restrictions, mainly regarding calibration.  We fixed some of these limitations in commits 687eeec8da8f840d5792a0de4d98c0d538cb881b, 6b4312b1a5a448caa652825eb979895bcc353f4f, and in a77877511fe3f11b06f1914ea4b5e4186e71071d itself. Issue #86 deals with the remaining general limitations of *HydPy* in supporting keyword-based parameter calibration.


[musk_mct](https://hydpy-dev.github.io/hydpy/master/musk_mct.html), our implementation of [Muskingum-Cunge method modified by Todini (MCT)](https://hess.copernicus.org/articles/11/1645/2007/] implemented in commit 52d48d4bfd3b3370df1bcf616b19e88a2c3f78c3, is also already in master.  It seems to work very similar to the original implementation of Todini.  However, there are three things that (maybe) need to be improved:

1. Maybe, there is a better way of [preventing negative discharge values](https://hydpy-dev.github.io/hydpy/master/musk_mct.html#no-baseflow)? 
2. It would be nice if one could select other channel geometries than only a trapezoidal profile.
3. The default error tolerance of the Pegasus-Iteration (for determining the reference water level) does not depend on the catchment size is way too low in most cases.

The second point calls for another modularisation effort. However, it is not urgent and requires more general discussions ([lstream_v001](https://hydpy-dev.github.io/hydpy/master/lstream_v001.html) only works for a triple trapezoid profile), so we postpone to a separate issue we will open later.

**Comment by tyralla** [comment]
>     3. The default error tolerance of the Pegasus-Iteration (for determining the reference water level) does not depend on the catchment size is way too low in most cases.

Commit 2a367e4688e9479bf8073fb5cb8f5e42a20ae9fb solves (or at least improves) this issue.  [ToleranceDischarge](https://hydpy-dev.github.io/hydpy/master/musk.html#hydpy.models.musk.musk_solver.ToleranceDischarge) now prepares its own value (which defines the absolute error tolerance) based on the catchment area.  I selected the river Main to give the [base example](https://hydpy-dev.github.io/hydpy/master/musk_mct.html#base-example) some context.  My first impression is that the new approach is a good compromise between computational speed and numerical accuracy. However, we still need to examine this in some real projects in more detail.
